### PR TITLE
Add VIP in Baremetal Cluster

### PR DESCRIPTION
**Pull request by kashifest**
Added virtual IP functionality in baremetal cluster manager

Co-authored-by: Xenwar <anwar.hassen@est.tech>

Signed-off-by: kashifest <kashif.khan@est.tech>

### CODE REVIEW

File: api/v1alpha2/baremetalcluster_types.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ const (
 
 // BareMetalClusterSpec defines the desired state of BareMetalCluster.
 type BareMetalClusterSpec struct {
+	APIEndpoint string `json:"apiEndpoint,omitempty"`
```
Comment (Jaakko-Os) [inline api/v1alpha2/baremetalcluster_types.go]: BareMetalClusterStatus struct already contains APIEndpoints list, why not use it?
The BareMetalClusterSpec comment also implies the struct content.

File: baremetal/cluster_manager.go [hunk 2]
```diff
@@ -78,10 +87,32 @@ func (s *ClusterManager) UpdateConfiguration() error {
...
+	if err != nil {
+		log.Fatalf("Invalid Port %s", err)
+	}
+
 	return []capbm.APIEndpoint{
 		{
-			Host: "",
-			Port: 8080,
+			Host: ip,
+			Port: port,
```
Comment (Jaakko-Os) [inline baremetal/cluster_manager.go]: when using BareMetalClusterStatus struct as it is, host and port could be:

			Host: s.Cluster.Status.APIEndpoints[0].Host,
			Port: s.Cluster.Status.APIEndpoints[0].Port,

**Review by Jaakko-Os — CHANGES_REQUESTED** [review]
BareMetalClusterStatus struct already contains APIEndpoints list, why not use it

File: examples/controlplane/controlplane.yaml [hunk 3]
```diff
@@ -43,6 +43,45 @@ spec:
...
+            priority 101
+            advert_int 1
+            authentication {
+                auth_type PASS
+                auth_pass 1111
+            }
+            virtual_ipaddress {
+                192.168.111.249
+            }
+        }
```
Comment (maelk) [inline examples/controlplane/controlplane.yaml]: This is also needed in the two other control planes

File: baremetal/cluster_manager.go [hunk 4]
```diff
@@ -78,10 +87,32 @@ func (s *ClusterManager) UpdateConfiguration() error {
 
 // IP returns the cluster manager IP address
 func (s *ClusterManager) APIEndpoints() ([]capbm.APIEndpoint, error) {
+	//Get IP address from spec, which gets it from posted cr yaml
+	// Once IP is handled, consider setting the port
+	endPoint := s.BareMetalCluster.Spec.APIEndpoint
```
Comment (maelk) [inline baremetal/cluster_manager.go:95]: what if endpoint is empty ? Maybe remove omitempty in the struct definition ?

**Review by maelk — CHANGES_REQUESTED** [review]
Can you please split this in two ? the first PR for the apiendpoint handling, the second PR for the controlplane.yaml modifications so that it is easy to revert when we have a proper load-balancing solution in place (and do the modifications for the two other control plane nodes)

**Comment by maelk** [comment]
> BareMetalClusterStatus struct already contains APIEndpoints list, why not use it

they are using it for the status, but we need also to give it as a spec

**Pull request merged**